src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs

```diff
@@ -0,0 +1,363 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+#nullable enable
+
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using DurableTask.Core;
+using DurableTask.Core.Exceptions;
+using DurableTask.Core.History;
+using DurableTask.Core.Middleware;
+using Microsoft.Azure.WebJobs.Host.Executors;
+
+namespace Microsoft.Azure.WebJobs.Extensions.DurableTask
+{
+    internal class OutOfProcMiddleware
+    {
+        private readonly DurableTaskExtension extension;
+
+        public OutOfProcMiddleware(DurableTaskExtension extension)
+        {
+            this.extension = extension;
+        }
+
+        // The below private properties are just passthroughs to properties or methods defined in the extension class.
+        // They exist simply to make it easier to copy/paste logic from the old middleware defined there to this file.
+
+        private DurableTaskOptions Options => this.extension.Options;
+
+        private EndToEndTraceHelper TraceHelper => this.extension.TraceHelper;
+
+        private ILifeCycleNotificationHelper LifeCycleNotificationHelper => this.extension.LifeCycleNotificationHelper;
+
+        private IApplicationLifetimeWrapper HostLifetimeService => this.extension.HostLifetimeService;
+
+        /// <summary>
+        /// Durable Task Framework orchestration middleware that invokes an out-of-process orchestrator function.
+        /// </summary>
+        /// <param name="dispatchContext">This middleware context provided by the framework that contains information about the orchestration.</param>
+        /// <param name="next">The next middleware handler in the pipeline.</param>
+        /// <exception cref="InvalidOperationException">Thrown if there's an internal failure in the middleware.</exception>
+        /// <exception cref="SessionAbortedException">Thrown if there is a recoverable error in the Functions runtime that's expected to be handled gracefully.</exception>
+        public async Task CallOrchestratorAsync(DispatchMiddlewareContext dispatchContext, Func<Task> next)
+        {
+            OrchestrationRuntimeState? runtimeState = dispatchContext.GetProperty<OrchestrationRuntimeState>();
+            if (runtimeState == null)
+            {
+                // This is not an orchestration; skip.
+                await next();
+                return;
+            }
```

Comment:
Would we find ourselves skipping here if we're dealing with an entity? In any case, I think it would be great to document why it's safe to "skip" in here, instead of error'ing out